api/envoy/api/v2/cds.proto

```diff
@@ -234,6 +234,32 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // UpstreamTransportSocket specifies the transport socket config used for upstream connection.
+  message UpstreamTransportSocket {
+    oneof socket_config {
+      core.TransportSocket transport_socket = 1;
+      auth.UpstreamTlsContext tls_context = 2;
+    }
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  // This should only be used for EDS typed cluster.
+  // The value of LbEndpoint.Metadata configuration under the *envoy.transport_socket_matcher*
+  // is used to decide which transport socket to use when connecting to that particular endpoint.
+  // For example, an endpoint with metadata
+  //
+  // .. code-block:: json
+  //
+  //   {
+  //     "envoy.transport_socket_matcher_label":
+  //        { "label": "mtlsEnabled" }
+  //   }
+  //
+  // This endpoint will use the transport_socket_matcher["mtlsEnabled"] configuration.
+  // In case of of label is missing or the transport_socket_matcher does not have corresponding
+  // entry in the map, Envoy fallbacks to the *tls_context* or *transport_socket* in the cluster.
+  map<string, UpstreamTransportSocket> transport_socket_matcher = 41;
```

Comment:
Can we just put this in EDS itself? I think I'm missing the design discussion here. Also, how does this relate to the aggregate cluster feature?